File: src/main/java/com/botequim/witchcraft/ai/minimax.java
package com.botequim.witchcraft.ai;

import scala.collection.Seq;
import scala.collection.JavaConversions;

abstract class MinimaxAlgorithm<Node> {
    abstract public boolean isTerminal(Node n);
    abstract public boolean victoryOrDraw(Node n, boolean player);
    abstract public double eval(Node n,  boolean player);
    abstract public Seq<Node> children(Node n, boolean player);
    public double max(Node node, boolean player, int depth) {
        if(depth <= 0)
            throw new UnsupportedOperationException("Cannot calc max with depth <= 0");
        //                System.out.println("Depth: " + depth);
        double maxEval = Double.NEGATIVE_INFINITY;
        for(Node c : JavaConversions.asJavaIterable(children(node, !player))) {
            //            System.out.println("Max Node: " + c);
            double newEval = min(c, player, depth -1);
            //              System.out.println("Eval: " + newEval);
            if(newEval > maxEval) {
                maxEval = newEval;
            }
        }
        return maxEval;
    }
    public double min(Node node, boolean player, int depth) {
        double minEval = Double.POSITIVE_INFINITY;
        double newEval;
        for(Node c : JavaConversions.asJavaIterable(children(node, player))) {
            //            System.out.println("Min Node: " + c);
            if(isTerminal(c)) {
                if(victoryOrDraw(c, player)) continue; // Max infinity so, never picked
                else return eval(c, player); // Defeat is the minimum possible value
            }
            if(depth == 0) newEval = eval(c, player);
            else newEval = max(c, player, depth);
            //            System.out.println("Eval: " + newEval);
            if(newEval < minEval) {
                minEval = newEval;
            }
        }
        return minEval;
    }
}

abstract class AlphaBetaMinimax<Node> {
    abstract public boolean isTerminal(Node n);
    abstract public boolean victoryOrDraw(Node n, boolean player);
    abstract public double eval(Node n,  boolean player);
    abstract public Seq<Node> children(Node n, boolean player);
    public double max(Node node, boolean player, int depth, double alpha, double beta) {
        double a = alpha;
        double b = beta;
        //        System.out.println("Depth: " + depth);
        double maxEval = Double.NEGATIVE_INFINITY;
        double newEval;
        for(Node c : JavaConversions.asJavaIterable(children(node, !player))) {
            if(isTerminal(c)) {
                if(victoryOrDraw(c, player)) return eval(c, player); // Best case
                else continue; // worst case
            }
            if(depth == 0) newEval = eval(c, player);
            else newEval = min(c, player, depth - 1, a, b);
            if(newEval > maxEval) {
                maxEval = newEval;
            }
            if(newEval >= b) return newEval;
            if(newEval > a) a = newEval;
        }
        return maxEval;
    }
    public double min(Node node, boolean player, int depth, double alpha, double beta) {
        double a = alpha;
        double b = beta;
        double minEval = Double.POSITIVE_INFINITY;
        double newEval;
        for(Node c : JavaConversions.asJavaIterable(children(node, player))) {
            if(isTerminal(c)) {
                if(victoryOrDraw(c, player)) continue; // Infinity so, never picked
                else return eval(c, player); // Defeat is the minimum possible value
            }
            if(depth == 0) newEval = eval(c, player);
            else newEval = max(c, player, depth - 1, a, b);
            if(newEval < minEval) {
                minEval = newEval;
            }
            if(newEval <= a) return newEval;
            if(newEval <= b) b = newEval;
        }
        return minEval;
    }
}

File: src/main/scala/com/botequim/game1/ai/minimax.scala
/*
 * Copyright 2013 Marcelo de Sena Lacerda
 *
 * This file is part of Botequim's Game 1.
 *
 * Game 1 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Game 1 is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Game 1.  If not, see <http://www.gnu.org/licenses/>.
 */

package org.botequim.game1.ai
import org.botequim._
import ai._
import game1.rules.{Game, Form}

class MinimaxAI extends MinimaxAlgorithm[Game] with AI {
  def isTerminal(node: Node): Boolean =
    node.gamePoints(true) == 0d ||
    node.gamePoints(false) == 0d
  def victoryOrDraw(n: Node, player: Boolean) =
    n.gamePoints(!player) == 0d

  def eval(n: Node, player: Boolean): Double =
    n.gamePoints(player) - n.gamePoints(!player) + n.spells(player).level * 0.01

  override def children(n: Node, player: Boolean) =
    super.children(n, player) map { c => c.getAftermath.getOrElse(c) }

  def getMove(sx: Seq[Node], player: Boolean, depth: Int): Node =
    children(sx.head, player) maxBy { n => max(n, player, depth) }

  def getMove(sx: Seq[Node], player: Boolean): Node = null
}

class MinimaxAlphaBetaAI extends AlphaBetaMinimax[Game] with AI {
  def isTerminal(node: Node): Boolean =
    node.gamePoints(true) == 0d ||
    node.gamePoints(false) == 0d
  def victoryOrDraw(n: Node, player: Boolean) =
    n.gamePoints(!player) == 0d

  def eval(n: Node, player: Boolean): Double =
    n.gamePoints(player) - n.gamePoints(!player) + n.spells(player).level * 0.01

  override def children(n: Node, player: Boolean) =
    super.children(n, player) map { c => c.getAftermath.getOrElse(c) }

  def getMove(sx: Seq[Node], player: Boolean, depth: Int): Node =
    children(sx.head, player) minBy { n => min(n, !player, depth, Double.NegativeInfinity, Double.PositiveInfinity) }

  def getMove(sx: Seq[Node], player: Boolean): Node = getMove(sx, player, 1)
}